[[ Bug 19032 ]] Sort out object deletion

livecodeali opened the pull request.

livecodeali committed:
[[ Bug 19032 ]] Sort out object deletion

peter-b commented on engine/src/control.cpp:
```diff
@@ -454,6 +454,59 @@ void MCControl::deselect()
+            if (!p_is_child)
+            {
+                MCCard *cptr = parent.GetAs<MCCard>();
+                cptr->removecontrol(this, False, True);
```
This is a pretty sweet-looking refactor. 💯 

I suggest `parent.GetAs<MCCard>()->removecontrol(this, False, True);` but YMMV.

livecodeali replied:
👍 

Not sure if this patch is quite correct yet anyway!

livecodeali committed:
[[ Cleanup ]] Remove unnecessary temp variables

livecodeali committed:
[[ Bug 18977 ]] Don't short-cut card setting in no ui mode
This was preventing openCard messages in no ui mode.

It also causes `MCCard::open` to be run when the only card on
a stack is deleted, in no ui mode. This causes any existing
controls in a card's `objptr`s to be reparented to the card,
causing a crash if a shared group is present, as this needs to
stay reparented to the stack.

livecodeali committed:
[[ Bug 18977 ]] Add test

livecodeali committed:
[[ Bug 18977 ]] [[ Bug 19032 ]] Add release notes

livecodeali commented:
I can't seem to write a test that reproduces 19032. We can't replicate the original recipe in the test system because currently data grids require user-interaction when deleted.

peter-b commented on engine/src/control.cpp:
```diff
@@ -454,6 +454,60 @@ void MCControl::deselect()
+        default:
+        {
+            parent.GetAs<MCStack>()->removecontrol(this);
+        }
```
I feel like this should be:

```C++
case CT_STACK:
    parent.GetAs<MCStack>()->removecontrol(this);
    break;
default:
    MCUnreachable();
```

This is because you're switching on the parents object type, and it makes sense to have a stack case for stacks. :)

livecodeali commented:
As this is pretty invasive, here is a patch which prevents the crash with much less churn.
https://github.com/livecode/livecode/pull/5071

livecodeali commented:
@peter-b Thanks, I have incorporated both your observations in the replacement PR!
